src/coreclr/vm/pgo.cpp

```diff
@@ -872,13 +888,150 @@ HRESULT PgoManager::getPgoInstrumentationResults(MethodDesc* pMD, SArray<ICorJit
+        {
+            BYTE* pEightByteData = (BYTE*)&dataItem;
+            for (int i = 0; i < 8; i++)
+            {
+                instrumentationData.Append(pEightByteData[i]);
+            }
+        }
+
+        return true;
+    }
+};
+
+HRESULT PgoManager::getPgoInstrumentationResultsFromR2RFormat(ReadyToRunInfo *pReadyToRunInfo,
+                                                              Module* pModule,
+                                                              PEDecoder* pNativeImage,
+                                                              BYTE* pR2RFormatData,
+                                                              size_t pR2RFormatDataMaxSize,
+                                                              BYTE** pAllocatedData,
+                                                              ICorJitInfo::PgoInstrumentationSchema** ppSchema,
+                                                              UINT32 *pCountSchemaItems,
+                                                              BYTE**pInstrumentationData)
+{
+    *pAllocatedData = NULL;
+    *ppSchema = NULL;
+    *pInstrumentationData = NULL;
+    *pCountSchemaItems = 0;
+
+    R2RInstrumentationDataReader r2rReader(pReadyToRunInfo, pModule, pNativeImage);
+
+    if (!ReadInstrumentationData(pR2RFormatData, pR2RFormatDataMaxSize, r2rReader))
+    {
+        return E_NOTIMPL;
+    }
+    else
+    {
+        while (r2rReader.instrumentationData.GetCount() & (sizeof(int64_t) - 1))
+        {
+            r2rReader.instrumentationData.Append(0);
+        }
+        size_t schemaDataSize = r2rReader.schemaArray.GetCount() * sizeof(ICorJitInfo::PgoInstrumentationSchema);
+        NewArrayHolder<BYTE> allocatedData = new BYTE[r2rReader.instrumentationData.GetCount() + schemaDataSize];
+
+        *pInstrumentationData = (BYTE*)allocatedData;
+        *pCountSchemaItems = r2rReader.schemaArray.GetCount();
+        memcpy(allocatedData, r2rReader.instrumentationData.OpenRawBuffer(), r2rReader.instrumentationData.GetCount());
+        r2rReader.instrumentationData.CloseRawBuffer();
+        *ppSchema = (ICorJitInfo::PgoInstrumentationSchema*)(((BYTE*)allocatedData) + r2rReader.instrumentationData.GetCount());
+        memcpy(*ppSchema, r2rReader.schemaArray.OpenRawBuffer(), schemaDataSize);
+        r2rReader.schemaArray.CloseRawBuffer();
+
+        allocatedData.SuppressRelease();
+        *pAllocatedData = allocatedData;
+        return S_OK;
+    }
+}
 #endif // DACCESS_COMPILE
 
-HRESULT PgoManager::getPgoInstrumentationResultsInstance(MethodDesc* pMD, SArray<ICorJitInfo::PgoInstrumentationSchema>* pSchema, BYTE**pInstrumentationData)
+HRESULT PgoManager::getPgoInstrumentationResultsInstance(MethodDesc* pMD, BYTE** pAllocatedData, ICorJitInfo::PgoInstrumentationSchema** ppSchema, UINT32 *pCountSchemaItems, BYTE**pInstrumentationData)
```

Comment:
Should this prefer "live" data over R2R data?

For instance, will R2R assemblies loaded under ZapDisable lookup embedded profile data for their methods even though they may have gone through Tier0?

Reply:
Umm, sure. I feel that ideally we would merge the data, but I don't really want to write that code right now, especially as it would be a duplication of the managed merger algorithm. Now that I think about it a bit, if we don't prefer in memory data we'll be losing all the potential for tiered pgo improving over static pgo, so... yeah, I'll swap it around.